src/jail_run.c

```diff
@@ -0,0 +1,282 @@
+    }
+
+    /* D-Bus names and interfaces, session bus */
+    if (jail_ptrv_length(dbus_user->own) ||
+        jail_ptrv_length(dbus_user->talk)) {
+        g_ptr_array_add(args, (gpointer) FIREJAIL_DBUS_USER_FILTER);
+        if (gutil_log_default.level > GLOG_LEVEL_DEBUG) {
+            g_ptr_array_add(args, (gpointer) FIREJAIL_DBUS_USER_LOG);
+        }
+        jail_run_add_dbus_opts(args, args_alloc, dbus_user,
+            FIREJAIL_DBUS_USER_OWN,
+            FIREJAIL_DBUS_USER_SEE,
+            FIREJAIL_DBUS_USER_TALK);
+    }
+
+    /* D-Bus names and interfaces, system bus */
+    if (jail_ptrv_length(dbus_system->own) ||
+        jail_ptrv_length(dbus_system->talk)) {
+        g_ptr_array_add(args, (gpointer) FIREJAIL_DBUS_SYSTEM_FILTER);
+        if (gutil_log_default.level > GLOG_LEVEL_DEBUG) {
+            g_ptr_array_add(args, (gpointer) FIREJAIL_DBUS_SYSTEM_LOG);
+        }
+        jail_run_add_dbus_opts(args, args_alloc, dbus_system,
+            FIREJAIL_DBUS_SYSTEM_OWN,
+            FIREJAIL_DBUS_SYSTEM_SEE,
+            FIREJAIL_DBUS_SYSTEM_TALK);
+    }
+
+    /* 3. Done with firejail options */
+    g_ptr_array_add(args, (gpointer) FIREJAIL_FINISH_OPT);
+
+    /* Append the program name and its arguments */
+    for (i = 0; i < argc; i++) {
+        g_ptr_array_add(args, argv[i]);
+    }
+
+    /* 4. NULL-terminate the array */
+    g_ptr_array_add(args, NULL);
+
+    /* Dump the full command line if debug log is enabled */
+#ifdef GUTIL_LOG_DEBUG
+    if (GLOG_ENABLED(GLOG_LEVEL_DEBUG)) {
+        const char* const* ptr = (const char**) args->pdata;
+        GString* tmp = g_string_new(*ptr++);
+
+        while (*ptr) g_string_append_printf(tmp, " %s", *ptr++);
+        GDEBUG("%s", tmp->str);
+        g_string_free(tmp, TRUE);
+    }
+#endif /* GUTIL_LOG_DEBUG */
+
+    /* Handle special privileges */
+    for (perm_ptr = rules->permits; *perm_ptr; perm_ptr++) {
+        if ((*perm_ptr)->type == JAIL_PERMIT_PRIVILEGED) {
+            const char* group = PRIVILEGED_GROUP;
+            const struct group* gr = getgrnam(group);
+
+            if (gr) {
+                egid = gr->gr_gid;
+                GDEBUG("Setting effective GID to %s (%u)", group, (uint)egid);
+            } else {
+                GWARN("Group '%s' is missing", group);
+            }
+        }
+    }
+
+    setfsuid(user->euid);
+    setfsgid(egid);
+
+    if (setgroups(user->ngroups, user->groups)) {
+        g_propagate_error(error, g_error_new(G_UNIX_ERROR, errno,
+            "setgroups error: %s", strerror(errno)));
+    } else if (setgid(egid)) {
+        g_propagate_error(error, g_error_new(G_UNIX_ERROR, errno,
+            "setgid(%u) error: %s", (guint)egid, strerror(errno)));
+    } else if (setuid(user->euid)) {
+        g_propagate_error(error, g_error_new(G_UNIX_ERROR, errno,
+            "setuid(%u) error: %s", (guint)user->euid, strerror(errno)));
+    } else {
+        execvp(conf->exec, (char**) args->pdata);
```

Comment:
does saved uid/gid survive exec()? should setresuid() & co be used to make sure it is not possible?

Reply:
Yes, good idea, I'm doing that and getting rid of libdbusaccess dependency in the process...

Reply:
Switched to setresuid, dropped libdbusaccess dependency